src/main/java/org/team1540/robot2025/subsystems/drive/ModuleIOTalonFX.java

```diff
@@ -0,0 +1,255 @@
+                turnStatorCurrent,
+                turnTemp);
+        ParentDevice.optimizeBusUtilizationForAll(drive, turn, cancoder);
+    }
+
+    @Override
+    public void updateInputs(ModuleIOInputs inputs) {
+        // Refresh all signals
+        StatusCode driveStatus = BaseStatusSignal.refreshAll(
+                drivePosition, driveVelocity, driveAppliedVolts, driveStatorCurrent, driveTemp);
+        StatusCode turnStatus =
+                BaseStatusSignal.refreshAll(turnPosition, turnVelocity, turnAppliedVolts, turnStatorCurrent, driveTemp);
+        StatusCode turnEncoderStatus = BaseStatusSignal.refreshAll(turnAbsolutePosition);
+
+        // Update drive inputs
+        inputs.driveConnected = driveConnectedDebounce.calculate(driveStatus.isOK());
+        inputs.drivePositionRads = Units.rotationsToRadians(drivePosition.getValueAsDouble());
+        inputs.driveVelocityRadPerSec = Units.rotationsToRadians(driveVelocity.getValueAsDouble());
+        inputs.driveAppliedVolts = driveAppliedVolts.getValueAsDouble();
+        inputs.driveSupplyCurrentAmps = driveSupplyCurrent.getValueAsDouble();
+        inputs.driveStatorCurrentAmps = driveStatorCurrent.getValueAsDouble();
+        inputs.driveTempCelsius = driveTemp.getValueAsDouble();
+
+        // Update turn inputs
+        inputs.turnConnected = turnConnectedDebounce.calculate(turnStatus.isOK());
+        inputs.turnEncoderConnected = turnEncoderConnectedDebounce.calculate(turnEncoderStatus.isOK());
+        inputs.turnAbsolutePosition = Rotation2d.fromRotations(turnAbsolutePosition.getValueAsDouble());
+        inputs.turnPosition = Rotation2d.fromRotations(turnPosition.getValueAsDouble());
+        inputs.turnVelocityRadPerSec = Units.rotationsToRadians(turnVelocity.getValueAsDouble());
+        inputs.turnAppliedVolts = turnAppliedVolts.getValueAsDouble();
+        inputs.turnSupplyCurrentAmps = turnSupplyCurrent.getValueAsDouble();
+        inputs.turnStatorCurrentAmps = turnStatorCurrent.getValueAsDouble();
+        inputs.turnTempCelsius = turnTemp.getValueAsDouble();
+
+        // Update odometry inputs
+        inputs.odometryTimestamps =
+                timestampQueue.stream().mapToDouble(Double::doubleValue).toArray();
+        inputs.odometryDrivePositionsRads = drivePositionQueue.stream()
+                .mapToDouble(Units::rotationsToRadians)
+                .toArray();
+        inputs.odometryTurnPositions =
+                turnPositionQueue.stream().map(Rotation2d::fromRotations).toArray(Rotation2d[]::new);
+        timestampQueue.clear();
+        drivePositionQueue.clear();
+        turnPositionQueue.clear();
+    }
+
+    @Override
+    public void setDriveOpenLoop(double input) {
+        drive.setControl(
+                switch (constants.DriveMotorClosedLoopOutput) {
+                    case Voltage -> driveVoltageReq.withOutput(input);
+                    case TorqueCurrentFOC -> driveTorqueReq.withOutput(input);
+                });
+    }
```

Comment:
if I say open loop don't I mean open loop? what are you doing here

Reply:
Ok this was taken from the AdvantageKit swerve template. What's happening here is the the open loop mode depends on the output mode of the motor in closed loop, so if it runs closed loop voltage control then it runs open loop voltage control here, and if it runs closed loop current control it runs open loop current control here.
I'd assume that this depends on the closed loop output mode because the `setOpenLoop` function is only called during characterization, where the output mode needs to be consistent with the closed loop mode in order to produce the correct gains.